src/org/omegat/util/StaticUtils.java

```diff
@@ -135,6 +138,20 @@ public static boolean isKey(KeyEvent e, int code, int modifiers) {
         return graphics.getAvailableFontFamilyNames();
     }
 
+    /**
+     * Returns the names of all font which can display emoji.
+     * @return Set of names.
+     */
+    public static String[] getUnicodeFontNames() {
```

Comment:
There are several issues:

1. This is incredibly slow. It takes 33 seconds on my machine (MacBook Pro 13" 2020, 2 GHz Quad-Core Intel Core i5)
2. This doesn't give meaningful results

On my machine there are three fonts that explicitly contain glyphs for U+1F603 😃:

- Apple Color Emoji
- BabelStone Han
- HanaMinA

The only one that *actually* supports emoji in the sense that users would expect is Apple Color Emoji. However Java Swing does not support Apple Color Emoji and probably never will.

It also doesn't make sense to choose Apple Color Emoji as your editor's font. That's just not how text rendering is intended to work in the modern era; instead the text rendering engine is supposed to appropriately fall back to appropriate fonts on a grapheme cluster level. As I understand it, it is a limitation on Java Swing that prevents this from working correctly.

The other fonts contain text (black & white outline) representations of U+1F603 😃 and, I suppose if you had very strange taste, could be appropriate for use as the editor font. However they don't and never will actually support emoji as users would expect: Java Swing is still incapable of displaying ZWJ-sequence emoji and other composed emoji like region flags.

I think the concept behind this PR (filtering the font list to enable the user to choose a font that will allow emoji support) is barking up the wrong tree. I think the ultimate goal of enabling better or easier emoji display is basically impossible while we are still using Java Swing (maybe SWT does better).

In my opinion the best we can do is the Aggressive Font Fallback feature. This basically applies your logic ("find a font that can show the character") to every character in the editor, separately. This mimics modern font fallback within the constraints of Swing's rendering capabilities.

(Note: I could be wrong about any of the above; if so I welcome corrections. Also the above might be specific to macOS; in that case it would be helpful to know where I am wrong with respect to other operating systems.)

Reply:
> The other fonts contain text (black & white outline) representations of U+1F603 smiley and, I suppose if you had very strange taste, could be appropriate for use as the editor font. However they don't and never will actually support emoji as users would expect: Java Swing is still incapable of displaying ZWJ-sequence emoji and other composed emoji like region flags.

I think Java Swing can display Unicode emoji-1.0 standard / Unicode 6 level of SMP characters, and I can display all emoji in a file https://sourceforge.net/p/omegat/bugs/_discuss/thread/e814c543af/66e8/attachment/emoicon.txt on OmegaT on my Mint Linux.

> This is incredibly slow. It takes 33 seconds on my machine (MacBook Pro 13" 2020, 2 GHz Quad-Core Intel Core i5)

Uum, core of the patch is code snippet here. It run a check loop for all installed fonts. 
https://sourceforge.net/p/omegat/bugs/1043/#9a7d/7438
Actually it is O(n) but a number of fonts may be a range of several hundreds to several thousand, and it seems no big issue. 
How do you have number of fonts? When I tested the patch on Linux, there are about 2600 of fonts, works in a second.
We can have a list of valid fonts for major platforms, macOS, Windows and Linux,  and we can check it whether installed or not. It would be not slow.

Reply:
> How do you have number of fonts?

The number of fonts inspected by the loop is 2,167.

I suspect that this is a difference between Linux and Mac: maybe Linux has some caching mechanism to describe which fonts contain glyphs for which codepoints, whereas on Mac each font has to actually be loaded into memory?

> We can have a list of valid fonts for major platforms

This is not a reasonable approach.

> I can display all emoji in a file https://sourceforge.net/p/omegat/bugs/_discuss/thread/e814c543af/66e8/attachment/emoicon.txt on OmegaT on my Mint Linux.

On Java 8 with Aggressive Font Fallback enabled and my editor font set to Dialog I get this:

<img width="257" alt="image" src="https://user-images.githubusercontent.com/2172537/123344831-a75f1c80-d58f-11eb-9dc4-e2431076ebae.png">

Note that whether a glyph can be displayed seems to have little to do with which version of Unicode the codepoint was assigned in.

Note also that with Aggressive Font Fallback I can enjoy my preferred editor font (in this case Piazzolla) and still see emoji:

<img width="304" alt="image" src="https://user-images.githubusercontent.com/2172537/123344929-df665f80-d58f-11eb-9b22-60a6a033ede8.png">

On Java 11 it seems that `Font#canDisplay` is broken: the Aggressive Font Fallback mechanism never kicks in, and your code reports that all of my fonts support emoji.

Here's what I get when I choose BabelStone Han as my editor font on Java 11:

<img width="272" alt="image" src="https://user-images.githubusercontent.com/2172537/123344510-f8224580-d58e-11eb-9957-58bad847abe0.png">

Reply:
> However they don't and never will actually support emoji as users would expect: Java Swing is still incapable of displaying ZWJ-sequence emoji and other composed emoji like region flags.

Yes, it is right.

Java 8 support Unicode 6.2 that does not include these sequence emojis,  but include simple emojis in the standard.
https://openjdk.java.net/jeps/133

Java 11 support Unicode 10 but UTS #51, Unicode Emoji, is out of scope of the JEP
https://openjdk.java.net/jeps/327

I think OmegaT can handle these characters in UI and editor when user has a  font.

Unicode 6.1：😀😗😙😑😮😯😴😛😕😟

OmegaT w/ Java 8  may not handle properly for following emojis even when user has a font.
OmegaT w/ Java 11 may show these emojis. 

Unicode 7.0：🙂🙁🕵🗣🕴🖕🖖🖐
Unicode 8.0：🤗🤓🤔🙄🤐🙃🤑🤒🤕🤖
Unicode 9.0：🤣🤠🤡🤥🤤🤢🤧🤴🤶🤵🤷
Unicode 10.0：🤩🤨🤯🤪🤬🤮🤫🤭🧐🧒

OmegaT can process other sequence of emoji/flags  as a tag if input filter is aware of it.

Reply:
`getAllFonts()`
>Returns an array **containing a one-point size instance** of all fonts available in this GraphicsEnvironment. Typical usage would be to allow a user to select a particular font. 

This is a reason it is slow on macOS.

Reply:
FYI: https://unicode.org/Public/emoji/13.1/ is a place where all Unicode 13.1 emoji test data(text file).